--- linux-nvidia-6.17-6.17.0.orig/drivers/acpi/arm64/ffh.c
+++ linux-nvidia-6.17-6.17.0/drivers/acpi/arm64/ffh.c
@@ -19,6 +19,9 @@
 				struct arm_smccc_1_2_regs *res);
 };
 
+static int (*ffh_custom_handler)(struct acpi_ffh_info *info,
+				 acpi_integer *value, void *region_context);
+
 int acpi_ffh_address_space_arch_setup(void *handler_ctxt, void **region_ctxt)
 {
 	enum arm_smccc_conduit conduit;
@@ -99,9 +102,38 @@
 			ffh_ctxt->invoke_ffh64_fn(r, r);
 			memcpy(value, r, ffh_ctxt->info.length);
 		}
+	} else if (ffh_custom_handler) {
+		int err = ffh_custom_handler(&ffh_ctxt->info, value,
+					     region_context);
+		if (err) {
+			pr_err("ARM FFH custom offset handler returned error=%d\n",
+			       err);
+			ret = AE_ERROR;
+		}
 	} else {
 		ret = AE_ERROR;
 	}
 
 	return ret;
 }
+
+int acpi_arm64_ffh_update_custom_offset_handler(
+	int (*handler)(struct acpi_ffh_info *info, acpi_integer *value,
+		       void *region_context))
+{
+	if (!handler) {
+		pr_debug("ARM FFH custom offset handler unregistered\n");
+		ffh_custom_handler = NULL;
+		return 0;
+	}
+
+	if (ffh_custom_handler)
+		pr_debug("ARM FFH custom offset handler updated\n");
+	else
+		pr_debug("ARM FFH custom offset handler registered\n");
+
+	ffh_custom_handler = handler;
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(acpi_arm64_ffh_update_custom_offset_handler);
--- linux-nvidia-6.17-6.17.0.orig/include/linux/acpi.h
+++ linux-nvidia-6.17-6.17.0/include/linux/acpi.h
@@ -1578,10 +1604,26 @@
 					     void **region_ctxt);
 extern int acpi_ffh_address_space_arch_handler(acpi_integer *value,
 					       void *region_context);
+int acpi_ffh_address_space_arch_update_custom_offset_handler(
+	int (*handler)(struct acpi_ffh_info *info, acpi_integer *value,
+		       void *region_context));
 #else
 static inline void acpi_init_ffh(void) { }
 #endif
 
+#if defined(CONFIG_ACPI_FFH) && defined(CONFIG_ARM64)
+int acpi_arm64_ffh_update_custom_offset_handler(
+	int (*handler)(struct acpi_ffh_info *info, acpi_integer *value,
+		       void *region_context));
+#else
+static inline int acpi_arm64_ffh_update_custom_offset_handler(
+	int (*handler)(struct acpi_ffh_info *info, acpi_integer *value,
+		       void *region_context))
+{
+	return -EOPNOTSUPP;
+}
+#endif
+
 #ifdef CONFIG_ACPI
 extern void acpi_device_notify(struct device *dev);
 extern void acpi_device_notify_remove(struct device *dev);
