--- a/drivers/usb/host/xhci.h
+++ b/drivers/usb/host/xhci.h
@@ -1644,6 +1644,7 @@
 #define XHCI_CDNS_SCTX_QUIRK	BIT_ULL(48)
 #define XHCI_ETRON_HOST	BIT_ULL(49)
 #define XHCI_LIMIT_ENDPOINT_INTERVAL_9 BIT_ULL(50)
+#define XHCI_NVIDIA_MT8901_HOST		BIT_ULL(51)
 
 	unsigned int		num_active_eps;
 	unsigned int		limit_active_eps;
--- a/drivers/usb/host/xhci.c
+++ b/drivers/usb/host/xhci.c
@@ -3702,6 +3702,11 @@
 	if (ret < 0)
 		goto cleanup;
 
+	if (xhci->quirks & XHCI_NVIDIA_MT8901_HOST) {
+		xhci_hub_control(hcd, SetPortFeature, USB_PORT_FEAT_U1_TIMEOUT, 0, NULL, 0);
+		xhci_hub_control(hcd, SetPortFeature, USB_PORT_FEAT_U2_TIMEOUT, 0, NULL, 0);
+	}
+
 	spin_lock_irqsave(&xhci->lock, flags);
 	for (i = 0; i < num_eps; i++) {
 		ep_index = xhci_get_endpoint_index(&eps[i]->desc);
--- a/drivers/usb/host/xhci-hub.c
+++ b/drivers/usb/host/xhci-hub.c
@@ -1290,6 +1290,17 @@
 			}
 			port_li = readl(port->addr + PORTLI);
 			status = xhci_get_ext_port_status(temp, port_li);
+
+			/*
+			 * In MT8901 USB host controller, the lane count can be wrongly set
+			 * to 2 instead of 1 for USB Gen2x1 devices due to a HW Bug. As a SW
+			 * WAR, check if port speed is 0x5 (SuperSpeedPlus Gen2x1) in
+			 * PORTSC register and update the lane count as 1.
+			 */
+			if ((xhci->quirks & XHCI_NVIDIA_MT8901_HOST) &&
+			    DEV_SUPERSPEEDPLUS(temp))
+				status &= ~0xff00;
+
 			put_unaligned_le32(status, &buf[4]);
 		}
 		break;
--- a/drivers/usb/host/xhci-plat.c
+++ b/drivers/usb/host/xhci-plat.c
@@ -266,6 +266,9 @@
 		if (device_property_read_bool(tmpdev, "xhci-skip-phy-init-quirk"))
 			xhci->quirks |= XHCI_SKIP_PHY_INIT;
 
+		if (device_property_read_bool(tmpdev, "xhci-nvidia-mediatek-host"))
+			xhci->quirks |= XHCI_NVIDIA_MT8901_HOST;
+
 		device_property_read_u32(tmpdev, "imod-interval-ns",
 					 &xhci->imod_interval);
 		device_property_read_u16(tmpdev, "num-hc-interrupters",
